WordPress/openverse-catalog
# Trigger TSV loading immediately after workflow

## Fixes


Fixes #285 by @AetherUnbound

## Description


At long last! This PR makes some significant changes to the TSV loading process.

**TL;DR**: The TSV loading steps are now connected within the provider workflow DAG and get triggered immediately after the provider finishes/fails.

**The long-winded description**:

The current TSV workflow is:
1. Provider workflow runs
1. TSV loader runs asynchronously every minute and ingests the oldest file

It is more desirable to have synchronous loading for the case where we're ingesting a resource daily/monthly for a few reasons:
- Running the TSV loader every minute results in a very large number of skipped runs (currently), see #309.
- If a problem occurs with a sheet, that sheet is moved to a different location and we have to hunt down the errant file. This also prevents the file from being easily re-run once it's corrected as it must be moved back into the staging location.
- It makes it easier to determine which provider TSV caused the load to fail, since the steps are attached directly to the DAG.

The new change adds one or several [task groups](https://www.astronomer.io/guides/task-groups) on the end of our **basic provider workflow DAG**. I have intentionally left all pieces of the current TSV loader functioning for when we get to enabling the partitioned re-ingestion DAGs. As such, this PR only makes minor tweaks to existing functions; where differing functionality is needed, new functions have been added to support this work.

It's important to note that each provider workflow DAG must now report the media types it supports. For most of our DAGs this is `image`, but any provider script that produces media records beyond that must do so explicitly (see: audio DAGs). The setup is complicated _a bit_ by the fact that a provider script can produce multiple media types (see: Wikimedia commons). In order to handle this case, each media type generates its own ingestion task group. This also helps ensure that the loads are independent of one another.

As described above, the current method ingests the least recently modified file. This new method gets handed the file to ingest explicitly from upstream tasks via [XComs](https://airflow.apache.org/docs/apache-airflow/stable/concepts/xcoms.html). Due to the way our provider API scripts are set up, the location on disk must be retrieved from the provider API module itself. I've added a wrapper function in the DAG which accomplishes this. Ideally with the work from #229 and other efforts, we'll be able to construct the output path using the DAG's execution date and pass that into the `MediaStore` (rather than retrieving it from the store).

In addition to new tests, I've also updated some of the tests to factor out common fixtures.

Phew, okay, I think that's it. There's a lot going on in this PR, even though it isn't a _ton_ of code. Let me know if there's anything I should explain more thoroughly. Much of this is intermediate work to 1) help us get up and running now, 2) pave the way for future work, and 3) prevent this PR from being enormous!


### Screenshots

**Simple, single-media DAG**
![Screenshot_2022-02-10_11-32-54](https://user-images.githubusercontent.com/10214785/153487887-de4afc48-fd61-43c5-aa69-5918c83c489f.png)


**Multi-media DAG**
![Screenshot_2022-02-10_11-33-08](https://user-images.githubusercontent.com/10214785/153487897-a54c82c3-d0f5-4902-9522-a16bc0fa6787.png)

## Suggested review order
1. `dags/common/storage/media.py` & `dags/common/loader/paths.py`- making some attributes public so they can be used in the DAG
2. `dags/common/loader/*` - new, more generic functions for the TSV loading steps
3. `dags/providers/*_workflow.py` - added media types where necessary
4. `dags/common/dag_factory.py` - bulk of the changes, adding on the TSV loading steps to the end of the ingestion workflow
5. `tests/*` and everything else

## Testing Instructions

1. `just test`
1. Run a provider workflow
1. Either "mark successful" or "mark failed" in the UI after some lines have been written
1. Observe that the loader runs immediately after the workflow step and ingests data directly into the database

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
